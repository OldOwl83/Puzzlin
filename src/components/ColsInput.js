import { useContext, useState, useEffect } from "react";
import { PuzzleGrid } from "../RompeCocosApp";

export const ColsInput = () => {

    const { puzzleGrid, setPuzzleGrid } = useContext( PuzzleGrid );
    
    const [ input, setInput ] = useState( puzzleGrid[0] );
    
    useEffect(() => {
        
        if( input >= 3 || input == '' )
        {
            if( input >= 3 )
                setPuzzleGrid( [ Number( input ), puzzleGrid[1] ] );
            else
                setPuzzleGrid( [ Number( 3 ), puzzleGrid[1] ] );
        }else
        {
            setPuzzleGrid( [ Number( 3 ), puzzleGrid[1] ] );
            setInput( 3 );
        }
    }, [ input, setPuzzleGrid ]);

    return <input 
                type="number" 
                value={ input }
                min="3"
                onChange={ ( e ) => setInput( e.target.value) } 
            />
}
