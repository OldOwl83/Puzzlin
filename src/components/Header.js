import React from 'react';
import { Link, useLinkClickHandler } from 'react-router-dom';

export const Header = () => {

    const linker = useLinkClickHandler( '/', { 
        target: '_self',
        replace: false,
        state: {
            contScroll: 0,
            selectedPhoto: 'images/portada1.png',
            author: 'Mauro Donnantuoni',
        }
    });

    return (
        <>
        <Link to='/' id='toHomeLink' onClick={ linker }>
            <header>
                <h1>Puzzlin</h1>
                <div id="logoContainer">
                    <img src="images/rompecabezas.png" alt="logo" />
                </div>
            </header>
        </Link>

        <hr />
        </>
    );
};
