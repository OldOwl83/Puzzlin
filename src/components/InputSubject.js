/*
* An input box to set the search keywords in the URL query parameters.
*/

import React, { useState } from "react";
import { useLocation, useNavigate } from "react-router-dom";

import '../styles/InputSubject.css';

export const InputSubject = () => {

    // console.log( 'InputSubject' );

    const navigate = useNavigate();
    const location = useLocation();

    const [ input, setInput ] = useState( '' );
    
    const submitInput = (e) => {
        e.preventDefault();
        
        setInput( '' );

        navigate( `../?q=${ input }`, { 
            //if the query is repeated immediately, the location is not duplicated in browser history.
            replace: location.search === `?q=${ input }` ? true : false,
     } );
    }

    return (

        <h2>Enter a subject and select a photo. Then, press on the preview to start.
            <form onSubmit={ submitInput }>
                <input 
                    id="inputSubject"
                    type="text"
                    autoComplete="on"
                    value={ input }
                    onChange={ (e) => setInput( e.target.value ) }
                />
            </form>
        </h2>
    )
}