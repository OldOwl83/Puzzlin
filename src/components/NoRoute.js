import React from 'react';

export const NoRoute = () => {
  return <p className='message'>Nonexistent path.</p>;
};
