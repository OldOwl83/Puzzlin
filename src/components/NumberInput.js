/*
* General purpose number input. It receives by parameters a function which changes an external value or state, the minimum and maximum numbers allowed and the initial value. If the user tries to exceed the set range, it automatically sets the minimum or maximum accordingly. It sets the value as an integer data type.
*/

import { useState, useEffect, memo } from "react";

export const NumberInput = memo(( { setValue, min, max, initialValue }) => {

    const [ input, setInput ] = useState( initialValue );
    
    useEffect(() => {
    
        if( ( input >= min && input <= max ) || input === '' ) //It must include '' to allow erase manually the text in the input box.
        {
            if( input >= min && input <= max )
                setValue( Number( input ) );
            else 
                setValue( Number( min ) ); // If the input box remain empty, the set value is the minimum.
        }else if( input > max )
        {
            setValue( Number( max ) );
            setInput( max );
        }else
        {
            setValue( Number( min ) )
            setInput( min );
        }
    }, [ input, min, max, setValue ]);

    return (
    
        <input 
            type="number" 
            value={ input }
            min={ min }
            max={ max }
            onChange={ ( e ) => setInput( e.target.value ) } 
        />
    );
});
