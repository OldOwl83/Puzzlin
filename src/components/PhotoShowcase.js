import { useContext, useState, useEffect } from "react";
import { MainPhotoContext } from "../Puzzlin";

import '../styles/PhotoShowcase.css';

export const PhotoShowcase = ( { photos = [] } ) => {
    
    // console.log(photos);

    const { setPhotoPuzzle } = useContext( MainPhotoContext );

    const [ sampler, setSampler ] = useState( photos );

    const [ photosContDisplacement, setPhotosContDisplacement ] = useState( 0 );
    const [ photosContHeight, setPhotosContHeight ] = useState( 0 );

    useEffect(() => {
        
        if( photos.length > 0)
            setSampler( photos );
        else if( sessionStorage.getItem( 'previousPhotos' ))
            setSampler( JSON.parse( sessionStorage.getItem( 'previousPhotos' ) ) );
        else
            setSampler( [] );

    }, [ setSampler, photos ]);

    useEffect( () => {
        setPhotosContHeight( sampler.length * 110 );
        setPhotosContDisplacement( 0 );

    }, [ sampler ]);

    return (
        <div id="sampler">
            <div 
                className="material-icons arrows"
                onClick={ () => {
                    ( photosContDisplacement + 330 <= 0 ) ? setPhotosContDisplacement( photosContDisplacement + 330 ) : setPhotosContDisplacement( 0 );
                } }    
            >keyboard_double_arrow_up</div>

            <div id="photosHidder">
                <div id="photosContainer" style={ { top: photosContDisplacement, height: photosContHeight } }>
                    
                    { sampler.map( ( { id, description, urlSmall } ) => {
                    
                            return (

                                <div key={ id } className="photos">

                                    <img className="animate__animated animate__zoomIn" 
                                        src={ urlSmall } 
                                        alt={ description } 
                                        onClick={ (e) => {
                                            setPhotoPuzzle( e.target.src );
                                        }}
                                    />
                                </div>)
                    })}
                    
                </div>
            </div>

            <div 
                className="material-icons arrows"
                onClick={ () => {
                    ( photosContDisplacement - 330 >= 440 - photosContHeight ) ? setPhotosContDisplacement( photosContDisplacement - 330 ) : setPhotosContDisplacement( 440 - photosContHeight );
                }}    
            >keyboard_double_arrow_down</div>
        </div>
    )
}
