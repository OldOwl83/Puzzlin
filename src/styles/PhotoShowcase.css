@import url(https://fonts.googleapis.com/icon?family=Material+Icons);

#sampler{
    
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: center;
    
    width: 17%;
    height: 520px;
    margin-top: 26px;
    padding: 10px;
    
    border: 2px solid gray;
    border-radius: 10px;
    box-shadow: 2px 2px 4px black;
    background-color: rgb(222, 223, 167);
    
}

#photosHidder{
    position: relative;
    
    width: 100%;
    height: 446px;

    overflow: hidden;
}

#photosContainer{
    position: absolute;

    display: flex;
    flex-direction: column;

    width: 100%;
    margin: auto;

    transition: top 0.5s ease-out;
}

.photos{
    width: 88%;
    height: 90px;

    margin: 10px auto;
    
}

#photosContainer img{
    width: 100%;
    height: 100%;
    
    object-fit: cover;
    
    border: 2px solid gray;
    border-radius: 10px;
    box-shadow: 3px 3px 4px black;
    
    transition: box-shadow 0.4s ease;
}

#photosContainer img:hover{
    box-shadow: none;
}

.arrows{
    font-weight: bold;
    color: #777;
    cursor: pointer;
}

.arrows:hover{
    color: #333;
}