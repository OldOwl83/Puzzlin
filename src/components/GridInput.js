/*
* It handles the contextual values of the puzzle grid. These values represent the columns and rows to cut out the puzzle photo and are in an array with two indices, which is a state belonging to the <Puzzlin /> component.
*/

import { useContext, useState, useEffect } from "react";

import { PuzzleGrid } from "../Puzzlin";
import { NumberInput } from "./NumberInput";

import "../styles/GridInput.css";

export const GridInput = () => {

    const { puzzleGrid, setPuzzleGrid } = useContext( PuzzleGrid );
    
    const [ cols, setCols ] = useState( puzzleGrid[0] );
    const [ rows, setRows ] = useState( puzzleGrid[1] );
    
    useEffect(() => {
        
        setPuzzleGrid( [ cols, rows ] );
        
    }, [ setPuzzleGrid, cols, rows ]);

    return (
    
        <div id="GridInput">
            <p>Columns
                <NumberInput setValue={ setCols } min={ 4 } max={ 8 } initialValue={ puzzleGrid[0] } />
                {/* It send de previous context values to initial value of this component, to maintain them across all routes. */}
            </p>
            <p>Rows
                <NumberInput setValue={ setRows } min={ 3 } max={ 8 } initialValue={ puzzleGrid[1] } />
            </p>
        </div>
    );
}
