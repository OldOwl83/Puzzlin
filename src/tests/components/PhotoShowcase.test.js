import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import { PhotoShowcase } from "../../components/PhotoShowcase";

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
    useLocation: () => { return { search: '?q=car' }; },
}));

describe('Tests on <PhotoShowcase /> component', () => {
  
    const photos = [ 
        {
            id: 63455,
            description: "Primera descripción", 
            urlRegular: 'www.primeraUrlRegular.com',
            urlSmall: 'www.primeraUrlSmall.com',
            author: 'José Antonio',
        },
        {
            id: 5555,
            description: "Segunda descripción", 
            urlRegular: 'www.segundaUrlRegular.com',
            urlSmall: 'www.segundaUrlSmall.com',
            author: 'Rodo Arruabarrena',
        },
        {
            id: 63415,
            description: "Primera descripción", 
            urlRegular: 'www.primeraUrlRegular.com',
            urlSmall: 'www.primeraUrlSmall.com',
            author: 'José Antonio',
        },
        {
            id: 5525,
            description: "Segunda descripción", 
            urlRegular: 'www.segundaUrlRegular.com',
            urlSmall: 'www.segundaUrlSmall.com',
            author: 'Rodo Arruabarrena',
        },
        {
            id: 63435,
            description: "Primera descripción", 
            urlRegular: 'www.primeraUrlRegular.com',
            urlSmall: 'www.primeraUrlSmall.com',
            author: 'José Antonio',
        },
        {
            id: 5554,
            description: "Segunda descripción", 
            urlRegular: 'www.segundaUrlRegular.com',
            urlSmall: 'www.segundaUrlSmall.com',
            author: 'Rodo Arruabarrena',
        },
        {
            id: 63555,
            description: "Primera descripción", 
            urlRegular: 'www.primeraUrlRegular.com',
            urlSmall: 'www.primeraUrlSmall.com',
            author: 'José Antonio',
        },
        {
            id: 5565,
            description: "Segunda descripción", 
            urlRegular: 'www.segundaUrlRegular.com',
            urlSmall: 'www.segundaUrlSmall.com',
            author: 'Rodo Arruabarrena',
        },
    ];

    const setPhotosContHeight = jest.fn();
    const setPhotosContScroll = jest.fn();
    
    const wrapper = mount( 
        <MemoryRouter>
            <PhotoShowcase photos={ photos } /> 
        </MemoryRouter>
    );

    test('It should match the snapshot and display two images.', () => {
      
        expect( wrapper ).toMatchSnapshot();

        expect( wrapper.find( 'img' ).length ).toBe( 8 );
    });
    
    test('The "navigate" function should be called with a new state location after clicking any images', () => {
        
        wrapper.find( 'img' ).at(0).simulate( 'click' );

        expect( mockNavigate ).toHaveBeenCalledTimes( 1 );
        expect( mockNavigate ).toHaveBeenCalledWith( { search: '?q=car' }, 
            {
                replace: true,
                state: {
                    contSroll: 0,
                    selectedPhoto: 'www.primeraUrlRegular.com',
                    author: 'José Antonio',
                }
            } );
    });

    test('The initial style of "photosContainer" element should be {top: 0; height: 220}', () => {
        
        expect( wrapper.find( '#photosContainer' ).prop( 'style' ).top).toBe( 0 );
        expect( wrapper.find( '#photosContainer' ).prop( 'style' ).height).toBe( wrapper.find( 'img' ).length * 110 );
    });

    test('The scroll of "photosContainer" element should change after clicking the arrow buttons', () => {
      
        wrapper.find( '.arrows' ).at(1).simulate( 'click' );

        //Don't work
        // expect( wrapper.find( '#photosContainer' ).prop( 'style' ).top).toBe( -330 );
    });
    
    
    
    

});
