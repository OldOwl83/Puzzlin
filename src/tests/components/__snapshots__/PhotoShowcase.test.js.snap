// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <PhotoShowcase /> component It should match the snapshot and display two images. 1`] = `
<div
  id="sampler"
>
  <div
    className="material-icons arrows"
    onClick={[Function]}
  >
    keyboard_double_arrow_up
  </div>
  <div
    id="photosHidder"
  >
    <div
      id="photosContainer"
      style={
        Object {
          "height": 880,
          "top": 0,
        }
      }
    >
      <div
        className="photos"
        key="63455"
      >
        <img
          alt="Primera descripción"
          className="animate__animated animate__backInLeft"
          draggable="false"
          onClick={[Function]}
          src="www.primeraUrlSmall.com"
        />
      </div>
      <div
        className="photos"
        key="5555"
      >
        <img
          alt="Segunda descripción"
          className="animate__animated animate__backInLeft"
          draggable="false"
          onClick={[Function]}
          src="www.segundaUrlSmall.com"
        />
      </div>
      <div
        className="photos"
        key="63415"
      >
        <img
          alt="Primera descripción"
          className="animate__animated animate__backInLeft"
          draggable="false"
          onClick={[Function]}
          src="www.primeraUrlSmall.com"
        />
      </div>
      <div
        className="photos"
        key="5525"
      >
        <img
          alt="Segunda descripción"
          className="animate__animated animate__backInLeft"
          draggable="false"
          onClick={[Function]}
          src="www.segundaUrlSmall.com"
        />
      </div>
      <div
        className="photos"
        key="63435"
      >
        <img
          alt="Primera descripción"
          className="animate__animated animate__backInLeft"
          draggable="false"
          onClick={[Function]}
          src="www.primeraUrlSmall.com"
        />
      </div>
      <div
        className="photos"
        key="5554"
      >
        <img
          alt="Segunda descripción"
          className="animate__animated animate__backInLeft"
          draggable="false"
          onClick={[Function]}
          src="www.segundaUrlSmall.com"
        />
      </div>
      <div
        className="photos"
        key="63555"
      >
        <img
          alt="Primera descripción"
          className="animate__animated animate__backInLeft"
          draggable="false"
          onClick={[Function]}
          src="www.primeraUrlSmall.com"
        />
      </div>
      <div
        className="photos"
        key="5565"
      >
        <img
          alt="Segunda descripción"
          className="animate__animated animate__backInLeft"
          draggable="false"
          onClick={[Function]}
          src="www.segundaUrlSmall.com"
        />
      </div>
    </div>
  </div>
  <div
    className="material-icons arrows"
    onClick={[Function]}
  >
    keyboard_double_arrow_down
  </div>
</div>
`;
